### ISSUE TITLE
Copying from the Chrome web browser doesn't work

### ISSUE THREAD

**Issue by joubertredrat** [synergy/synergy#5041]
Hi,

When I try to copy from browser on slave don't work, then I need to paste on notepad and copy from notepad to work. Both problem from server to slave.

Anyone have similar problem?

**Comment by nyetwurk** [comment]
Yes. I am seeing this as well, but only after a long time (server to client copy). If i restart the server the problem disappears (for a while)

**Comment by joubertredrat** [comment]
Hi, 
@nyetwurk true, now that you spoke, is similar here, isn't all time that problem appears

**Comment by nyetwurk** [comment]
The problem is in MSWindowsClipboard::get(). For some reason, after some time, the converter matcher code stops working (converter=null), but only for clipboards from GUIs. Clipboards from notepad (or cmd.exe) work fine. If you stop and restart the server, it works again.

**Comment by nyetwurk** [comment]
Looks like it is related to coming out of screen saver.
To reproduce, try setting the screen saver timeout on your server to 1 min.

I think I found the problem in isOwnedBySynergy() but I am still testing.

**Comment by nyetwurk** [comment]
I inserted this code:
````C++
    String s("converters:");
    for (ConverterList::const_iterator index = m_converters.begin();
                            index != m_converters.end(); ++index) {
        s+=synergy::string::sprintf(" %d,%d",
            (*index)->getFormat(),
            (*index)->getWin32Format());
    }
    LOG((CLOG_DEBUG "%s\n", s.c_str()));

    s = String("formats:");
    for (UINT win32Format = EnumClipboardFormats(0);
        win32Format != 0; win32Format = EnumClipboardFormats(win32Format)) {
        s+=synergy::string::sprintf(" %d", win32Format);
    }
    LOG((CLOG_DEBUG "%s\n", s.c_str()));
````

My converters and clipboard formats go from:
````
[2015-10-01T18:14:38] DEBUG: converters: 0,13 1,8 2,49375
[2015-10-01T18:14:38] DEBUG: formats: 49161 49466 49375 49851 49866 13 1 49784 49171 16 7
````
to
````
[2015-10-01T18:16:14] DEBUG: converters: 0,13 1,8 2,49375
[2015-10-01T18:16:14] DEBUG: formats: 49161
````

Any idea what happened to the rest of the formats?

In particular, 49375 is no longer reported by getWin32Format()

**Comment by nyetwurk** [comment]
The problem is that if synergy service starts the synergy server too quickly, and a client connects before it switches from 
````
[2015-10-01T21:14:47] DEBUG: desk Screen-saver window is 0x007a0668
````
to
````
[2015-10-01T21:14:47] DEBUG: switched to desk "Default"
````
... something goes very wrong with the clipboard.

If you delay letting clients connect, and you get
````
[2015-10-01T21:25:50] DEBUG: desk Default window is 0x00170744
[2015-10-01T21:25:50] DEBUG: switched to desk "Default"
````
before a client connects, the clipboard works correctly. I am unsure of the correct fix.

**Comment by nyetwurk** [comment]
Update: has nothing to do with the client connection. Has to do with coming out of screen saver, regardless of client.

**Comment by ctlajoie** [comment]
@nyetwurk What was the purpose of shutting down the server when the screensaver starts in the first place?

**Comment by nyetwurk** [comment]
@ctlajoie I don't think it was intentional. There is code to shut down synergy when the desktop changes (presumably from one virtual desktop to another), and the screen saver activation was (inadvertently?) triggering that.

**Comment by ctlajoie** [comment]
@nyetwurk At the risk of sounding completely ignorant, I have to admit I don't understand what you mean when you say that the server shuts down when switching from one virtual desktop to another. On my Linux (client in my case), I frequently switch between 3 virtual desktops, but synergy continues to work. Are you saying that if my server was on my Linux box then switching to another desktop would cause synergy to stop working until I switch back to my primary desktop?

**Comment by nyetwurk** [comment]
@ctlajoie I definitely don't pretend to be an expert; I'm pretty much flying blind and making stuff up as I go along by just reading existing code.. but from what I can tell, changing virtual desktops *on the server* causes synergys to be shutdown, but then automatically restarted (at least on windows). The bug is that if the server is started while the screensaver is the active desktop, the clipboard code doesn't seem to get initialized properly... and then when the screen saver deactivates, and a proper desktop becomes active, the "right" thing isn't happening (presumably if it did, the clipboard would start working properly again). This is all a guess. I could be entirely wrong..

**Comment by ctlajoie** [comment]
@nyetwurk Could this be related to the fact that synergy stops working (can't move cursor to client) when an application with administrator privs has focus? I'm running Windows 10 as my server - I'm not sure if it happens in other windows versions. If you haven't seen it before, it's easy to repro. Just run notepad as administrator and try to move your mouse to your other system while that notepad has focus.

I would like to try to debug this but I have been unable to produce a functional build on windows. Perhaps you would be willing to help me with that? I'll give you my email address if that's ok with you.

**Comment by nyetwurk** [comment]
@ctlajoie I cannot reproduce that problem (admin privs program).

Not sure how much help I can be producing a windows build, it is a bit tricky. I mostly followed the instructions...

**Comment by ctlajoie** [comment]
@nyetwurk Interresting that this problem only shows up for me...

My problem is not the build specifically, but running synergys after the build results in an access violation on line 398 of ArchTaskBarWindows.cpp. I have tried to debug it but I cannot find the cause. I also get the same error from synergyc.

**Comment by database9-445** [comment]
I can confirm that I also see the administrative privileges glitch as well as the clipboard glitch.

**Comment by nyetwurk** [comment]
Perhaps both are UAC related? I always disable UAC entirely.

**Comment by nyetwurk** [comment]
Unfortunately, I am still seeing the problem, even with my fix.

**Comment by nyetwurk** [comment]
Looks like if the server crashes (or exits) and the watchdog automatically restarts it while the screen saver is active, calls to EnumClipboardFormats(0) only return 49161 (DataObject) and never anything else, even if the screen saver is no longer active. Only restarting the server yet again (while screen saver is inactive) makes EnumClipboardFormats(0) work. Not sure what the solution is, since MSWindowsWatchdog doesn't know if the screen saver is active. Anybody have any suggestions?

**Comment by galkinvv** [comment]
While debugging similar issue when EnumClipboardFormats doesn't return all formats but only DataObject I came to a conclusion similar to what Ron Koerner found here:
https://github.com/synergy/synergy/issues/3153#issuecomment-58851628

In short: EnumClipboardFormats in some situations due to Windows/3rdparty software bug returns truncated list, but CountClipboardFormats and GetClipboardData works fine, so synergy can workaround this by a patch similar to proposed in a link above.

The problem was reproduced with a standalone application that works ok in non-elevated mode and returned truncated list in elevated the same way that synergy is:

```c++
#include <Windows.h>
#include <chrono>
#include <thread>
#include <iostream>

using namespace std;
int main(int argc, char* argv[])
{
	CoInitialize(nullptr);
	for (;;)
	{
		this_thread::sleep_for(chrono::seconds(1));
		OpenClipboard(nullptr);
		UINT format = 0;
		ostream& out = std::cout;
		out << "last error " << GetLastError() << " ";
		out << "Count = " << CountClipboardFormats() << ' ';
		for (;;)
		{
			format = EnumClipboardFormats(format);
			if (!format)
			{
				break;
			}
			
			char name_buffer[100] = {};
			const char* name = 0;
			switch (format)
			{
				case CF_TEXT:
					name = "CF_TEXT";
					break;
				case CF_OEMTEXT:
					name = "CF_OEMTEXT";
					break;
				case CF_LOCALE:
					name = "CF_LOCALE";
					break;
				case CF_UNICODETEXT:
					name = "CF_UNICODETEXT";
					break;
				default:
					GetClipboardFormatNameA(format, name_buffer, sizeof(name_buffer));
					name = name_buffer;
			}
			out << format << "(" << name << ") ";
		}
		out << " last error " << GetLastError() << "\n";
		CloseClipboard();
	}
}

```

I tried the following to fix such behaviour with but this doesn't help:
```c++
CoInitializeSecurity(nullptr, 0, nullptr, nullptr, RPC_C_AUTHN_LEVEL_NONE, RPC_C_IMP_LEVEL_DELEGATE, nullptr, EOAC_NONE, nullptr);
```
Unfortunately now I lost the environment where that bug appears and can't rerproduce it.

It seems that this problem is not related to elevation but using GetClipboardData without enumerating foramts may fix it too.

**Comment by nyetwurk** [comment]
Interesting. I wonder why the screen saver is affecting this... why would having the server start while the screen saver is active cause it to start in elevated mode, where having the server start otherwise does not? Or is it the same bug but triggered by two different conditions?

**Comment by nyetwurk** [comment]
From looking at PR #3241 it does indeed seem to be related to elevated mode. I have added the alternate get() above to my code to run if EnumClipboardFormats() fails... still testing, but it would be nice to get some guidance from the devs here...

**Comment by nyetwurk** [comment]
@galkinvv I have verified that your workaround works for the case where EnumClipboardFormats() is broken. I reverted MSWindowsDesks to the old behavior, and was able to reproduce the issue - starting the server while the screen saver is active definitely triggers the bug, and the workaround works! I still am confused if there are other elevation mode related issues, however.

**Comment by XinyuHou** [comment]
@joubertredrat 
This is maybe duplicated to #5004

Which browser do you use?

**Comment by XinyuHou** [comment]
Can any one test this nightly builds?

http://synergy-project.org/nightly?filter=nyetwurk-master-alpha-3eaefc

**Comment by joubertredrat** [comment]
I installed here now and I will test, I report here if have some problems on this issue

**Comment by nbolton** [comment]
@joubertredrat You didn't say which browser, so I'm assuming you mean Chrome.

**Comment by joubertredrat** [comment]
Hi guys,
Sorry @nbolton is in any browser.

@XinyuHou problem here persists, but it doesn't happen all the time, as happens, causes timeout on client, then I need to restart client to work again.

What better log options I need to set to put here?

**Comment by nbolton** [comment]
Ready for testing! :+1: 

http://synergy-project.org/nightly?filter=v1.7.6-rc1-c97bd2c

Thanks @XinyuHou! :smile:

**Comment by joubertredrat** [comment]
Thanks guys, I will test here now

**Comment by bsparrow435** [comment]
Confirmed clipboard working on win10 64-bit server with OSX10.11.1 client using chrome on server and terminal on Mac. Does not work on previous version.

**Comment by joubertredrat** [comment]
At this moment, working fine with Ubuntu 12.04 x64 as server and Windows 8 x64 as client.

**Comment by joubertredrat** [comment]
Hi guys,
Problem persists, in some moments if I copy from browser, client reports timeout from server and I need to restart server to sync again, then If I copy again works fine.

**Comment by joubertredrat** [comment]
Hi guys,

Again timeout, I changed my log level to debug1 to see if I discover why this fail with me

![synergy](https://cloud.githubusercontent.com/assets/1520407/11238178/173165a8-8dcb-11e5-80e3-aa76c3daee0b.jpg)

**Comment by joubertredrat** [comment]
Hi guys,

I don't know if this is relevant for this, but I think that can help.

https://gist.github.com/joubertredrat/3bc24bdf488891a926a6

![sy_1](https://cloud.githubusercontent.com/assets/1520407/11568488/84d5eaa2-99d3-11e5-84dd-a435083f0391.jpg)

**Comment by XinyuHou** [comment]
@joubertredrat 
Your issue is more like #4768

**Comment by joubertredrat** [comment]
Thanks @XinyuHou, I will move this to correct issue

**Comment by XinyuHou** [comment]
Hey guys,

Please try this nightly.
http://synergy-project.org/nightly?filter=synergy-v1.7.6-rc1-5287c62

**Comment by joubertredrat** [comment]
Hi @XinyuHou I have new informations about copy-paste and I don't know if is sufficient to open new issue.

Server: Ubuntu 16.04.1  x64 and Synergy 1.8.8-stable commit 25a8cb2
Slave: Windows 8.1 x64 and Synergy 1.8.8-stable commit 25a8cb2

On server I have [Sublime Text](https://www.sublimetext.com/) and [Atom](https://atom.io/), several times when I copy all text from Sublime Text and try paste on Windows, paste fails with empty data and I try a lot of times. If I copy text from Sublime Text to Atom on Server and copy from Atom and paste on client, works fine. Then I think that copy-paste can fail depending of application and I have similar problem with other applications.

What you think?

**Comment by XinyuHou** [comment]
Hi Joubert,
Thank you for the information.
It's possible that some applications can affect Synergy clipboard. It could also
be a bug in Synergy itself.
> If I copy text from Sublime Text to Atom on Server and copy from Atom and
paste on client, works fine.
Please try to copy directly from sublime to a client after you have successfully
copied something using the above approach. Make sure each time you copy
different contexts.
If the problem persists, it's likely to be an application specific issue.
Otherwise it's more likely to be a bug in Synergy itself.
Thanks,Jerry  





On Mon, May 29, 2017 2:00 PM, Joubert RedRat notifications@github.com  wrote:
Hi @XinyuHou  I have new informations about copy-paste and I don't know if is
sufficient to open new issue.

Server: Ubuntu 16.04.1 x64 and Synergy 1.8.8-stable commit 25a8cb2
Slave: Windows 8.1 x64 and Synergy 1.8.8-stable commit 25a8cb2

On server I have Sublime Text  and Atom, several times when I copy all text from
Sublime Text and try paste on Windows, paste fails with empty data and I try a
lot of times. If I copy text from Sublime Text to Atom on Server and copy from
Atom and paste on client, works fine. Then I think that copy-paste can fail
depending of application and I have similar problem with other applications.

What you think?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.



--
Jerry HouSenior Software EngineerSynergy Team
Symless
We're looking formarketing andengineering people in the UKsymless.com/careers


Symless is now hiring digital marketers and software engineers. We're building
exciting new software products to bring your devices together, making the world
more seamless. Symless is the company behind Synergy, keyboard and mouse sharing
software used by millions worldwide.
